{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cc9759ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>CHROMOSOME</th>\n",
       "      <th>POSITION</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs6771792</td>\n",
       "      <td>3</td>\n",
       "      <td>12276472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs11709077</td>\n",
       "      <td>3</td>\n",
       "      <td>12311507</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1801282</td>\n",
       "      <td>3</td>\n",
       "      <td>12368125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs1899951</td>\n",
       "      <td>3</td>\n",
       "      <td>12369840</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs4376068</td>\n",
       "      <td>3</td>\n",
       "      <td>186980329</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          SNP  CHROMOSOME   POSITION\n",
       "0   rs6771792           3   12276472\n",
       "1  rs11709077           3   12311507\n",
       "2   rs1801282           3   12368125\n",
       "3   rs1899951           3   12369840\n",
       "4   rs4376068           3  186980329"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Load your datasets\n",
    "gwas_df = pd.read_csv(r\"C:\\Users\\gargi\\Desktop\\Publication\\Transcriptomic_and_Epigenomic_Features\\genome_wide_significant_snps.csv\")\n",
    "\n",
    "gwas_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8bc97c66",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 236 entries, 0 to 235\n",
      "Data columns (total 3 columns):\n",
      " #   Column      Non-Null Count  Dtype \n",
      "---  ------      --------------  ----- \n",
      " 0   SNP         236 non-null    object\n",
      " 1   CHROMOSOME  236 non-null    int64 \n",
      " 2   POSITION    236 non-null    int64 \n",
      "dtypes: int64(2), object(1)\n",
      "memory usage: 5.7+ KB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "# Summary\n",
    "print(gwas_df.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7e52fb8a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>CHROMOSOME</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>RISK_ALLELE</th>\n",
       "      <th>OTHER_ALLELE</th>\n",
       "      <th>P_VALUE</th>\n",
       "      <th>OR</th>\n",
       "      <th>OR_95L</th>\n",
       "      <th>OR_95U</th>\n",
       "      <th>N_CASES</th>\n",
       "      <th>N_CONTROLS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs12565286</td>\n",
       "      <td>1</td>\n",
       "      <td>711153</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.16</td>\n",
       "      <td>1.11</td>\n",
       "      <td>0.96</td>\n",
       "      <td>1.30</td>\n",
       "      <td>4040</td>\n",
       "      <td>17875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs3094315</td>\n",
       "      <td>1</td>\n",
       "      <td>742429</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.82</td>\n",
       "      <td>1.01</td>\n",
       "      <td>0.95</td>\n",
       "      <td>1.07</td>\n",
       "      <td>6673</td>\n",
       "      <td>25149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs2905035</td>\n",
       "      <td>1</td>\n",
       "      <td>765522</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.36</td>\n",
       "      <td>1.04</td>\n",
       "      <td>0.96</td>\n",
       "      <td>1.11</td>\n",
       "      <td>4984</td>\n",
       "      <td>23960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs2980319</td>\n",
       "      <td>1</td>\n",
       "      <td>766985</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.39</td>\n",
       "      <td>1.03</td>\n",
       "      <td>0.97</td>\n",
       "      <td>1.09</td>\n",
       "      <td>7575</td>\n",
       "      <td>27012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs4040617</td>\n",
       "      <td>1</td>\n",
       "      <td>769185</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.50</td>\n",
       "      <td>1.03</td>\n",
       "      <td>0.95</td>\n",
       "      <td>1.11</td>\n",
       "      <td>4984</td>\n",
       "      <td>23960</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          SNP  CHROMOSOME  POSITION RISK_ALLELE OTHER_ALLELE  P_VALUE    OR  \\\n",
       "0  rs12565286           1    711153           G            C     0.16  1.11   \n",
       "1   rs3094315           1    742429           A            G     0.82  1.01   \n",
       "2   rs2905035           1    765522           G            A     0.36  1.04   \n",
       "3   rs2980319           1    766985           T            A     0.39  1.03   \n",
       "4   rs4040617           1    769185           A            G     0.50  1.03   \n",
       "\n",
       "   OR_95L  OR_95U  N_CASES  N_CONTROLS  \n",
       "0    0.96    1.30     4040       17875  \n",
       "1    0.95    1.07     6673       25149  \n",
       "2    0.96    1.11     4984       23960  \n",
       "3    0.97    1.09     7575       27012  \n",
       "4    0.95    1.11     4984       23960  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "diagram_df = pd.read_csv(r\"C:\\Users\\gargi\\Desktop\\Publication\\Transcriptomic_and_Epigenomic_Features\\DIAGRAMv3.2012DEC17.csv\")\n",
    "\n",
    "diagram_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "523ceac4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 2473441 entries, 0 to 2473440\n",
      "Data columns (total 11 columns):\n",
      " #   Column        Dtype  \n",
      "---  ------        -----  \n",
      " 0   SNP           object \n",
      " 1   CHROMOSOME    int64  \n",
      " 2   POSITION      int64  \n",
      " 3   RISK_ALLELE   object \n",
      " 4   OTHER_ALLELE  object \n",
      " 5   P_VALUE       float64\n",
      " 6   OR            float64\n",
      " 7   OR_95L        float64\n",
      " 8   OR_95U        float64\n",
      " 9   N_CASES       int64  \n",
      " 10  N_CONTROLS    int64  \n",
      "dtypes: float64(4), int64(4), object(3)\n",
      "memory usage: 207.6+ MB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "# Summary\n",
    "print(diagram_df.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "336eca23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>CHROMOSOME</th>\n",
       "      <th>POSITION</th>\n",
       "      <th>RISK_ALLELE</th>\n",
       "      <th>OTHER_ALLELE</th>\n",
       "      <th>P_VALUE</th>\n",
       "      <th>OR</th>\n",
       "      <th>OR_95L</th>\n",
       "      <th>OR_95U</th>\n",
       "      <th>N_CASES</th>\n",
       "      <th>N_CONTROLS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs12565286</td>\n",
       "      <td>1</td>\n",
       "      <td>711153</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.16</td>\n",
       "      <td>1.11</td>\n",
       "      <td>0.96</td>\n",
       "      <td>1.30</td>\n",
       "      <td>4040</td>\n",
       "      <td>17875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs3094315</td>\n",
       "      <td>1</td>\n",
       "      <td>742429</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.82</td>\n",
       "      <td>1.01</td>\n",
       "      <td>0.95</td>\n",
       "      <td>1.07</td>\n",
       "      <td>6673</td>\n",
       "      <td>25149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs2905035</td>\n",
       "      <td>1</td>\n",
       "      <td>765522</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.36</td>\n",
       "      <td>1.04</td>\n",
       "      <td>0.96</td>\n",
       "      <td>1.11</td>\n",
       "      <td>4984</td>\n",
       "      <td>23960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs2980319</td>\n",
       "      <td>1</td>\n",
       "      <td>766985</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.39</td>\n",
       "      <td>1.03</td>\n",
       "      <td>0.97</td>\n",
       "      <td>1.09</td>\n",
       "      <td>7575</td>\n",
       "      <td>27012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs4040617</td>\n",
       "      <td>1</td>\n",
       "      <td>769185</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.50</td>\n",
       "      <td>1.03</td>\n",
       "      <td>0.95</td>\n",
       "      <td>1.11</td>\n",
       "      <td>4984</td>\n",
       "      <td>23960</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          SNP  CHROMOSOME  POSITION RISK_ALLELE OTHER_ALLELE  P_VALUE    OR  \\\n",
       "0  rs12565286           1    711153           G            C     0.16  1.11   \n",
       "1   rs3094315           1    742429           A            G     0.82  1.01   \n",
       "2   rs2905035           1    765522           G            A     0.36  1.04   \n",
       "3   rs2980319           1    766985           T            A     0.39  1.03   \n",
       "4   rs4040617           1    769185           A            G     0.50  1.03   \n",
       "\n",
       "   OR_95L  OR_95U  N_CASES  N_CONTROLS  \n",
       "0    0.96    1.30     4040       17875  \n",
       "1    0.95    1.07     6673       25149  \n",
       "2    0.96    1.11     4984       23960  \n",
       "3    0.97    1.09     7575       27012  \n",
       "4    0.95    1.11     4984       23960  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation_df = pd.read_csv(r\"C:\\Users\\gargi\\Desktop\\Publication\\Transcriptomic_and_Epigenomic_Features\\bwMtrPidtktV3hgO.csv\")  \n",
    "\n",
    "annotation_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0dffa421",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 2473441 entries, 0 to 2473440\n",
      "Data columns (total 11 columns):\n",
      " #   Column        Dtype  \n",
      "---  ------        -----  \n",
      " 0   SNP           object \n",
      " 1   CHROMOSOME    int64  \n",
      " 2   POSITION      int64  \n",
      " 3   RISK_ALLELE   object \n",
      " 4   OTHER_ALLELE  object \n",
      " 5   P_VALUE       float64\n",
      " 6   OR            float64\n",
      " 7   OR_95L        float64\n",
      " 8   OR_95U        float64\n",
      " 9   N_CASES       int64  \n",
      " 10  N_CONTROLS    int64  \n",
      "dtypes: float64(4), int64(4), object(3)\n",
      "memory usage: 207.6+ MB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "# Summary\n",
    "print(annotation_df.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8c174bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Standardize rsID column names\n",
    "gwas_df.rename(columns={\"rsID\": \"Uploaded_variation\"}, inplace=True)\n",
    "diagram_df.rename(columns={\"SNP\": \"Uploaded_variation\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c157ee5",
   "metadata": {},
   "outputs": [],
   "source": [
    "diagram_df.rename(columns={\"SNP\": \"Uploaded_variation\"}, inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "806b1a6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge GWAS and DIAGRAM\n",
    "merged_1 = pd.merge(gwas_df, diagram_df, on=\"Uploaded_variation\", how=\"inner\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4851997",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge with functional annotations\n",
    "merged_final = pd.merge(merged_1, annotation_df, on=\"Uploaded_variation\", how=\"inner\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3285691",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop duplicates\n",
    "merged_final.drop_duplicates(inplace=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "067136c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter for genome-wide significance\n",
    "if 'P' in merged_final.columns:\n",
    "    merged_final = merged_final[merged_final['P'] < 5e-8]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ea4ca34",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save cleaned and merged data\n",
    "merged_final.to_csv(\"cleaned_merged_gwas_functional.csv\", index=False)\n",
    "\n",
    "# Preview result\n",
    "print(merged_final.shape)\n",
    "print(merged_final[['Uploaded_variation', 'SYMBOL', 'Consequence', 'P']].head())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
